flink-connector-netty/src/main/scala/org/apache/flink/streaming/connectors/netty/example/NettyUtil.scala

```diff
@@ -0,0 +1,160 @@
+    address.isLoopbackAddress match {
+      case true =>
+        // Address resolves to something like 127.0.1.1, which happens on Debian; try to find
+        // a better address using the local network interfaces
+        // getNetworkInterfaces returns ifs in reverse order compared to ifconfig output order
+        // on unix-like system. On windows, it returns in index order.
+        // It's more proper to pick ip address following system output order.
+        val activeNetworkIFs = NetworkInterface.getNetworkInterfaces.asScala.toSeq
+        val reOrderedNetworkIFs = SystemUtils.IS_OS_WINDOWS match {
+          case true => activeNetworkIFs
+          case false => activeNetworkIFs.reverse
+        }
+
+        reOrderedNetworkIFs.find { ni: NetworkInterface =>
+          val addr = ni.getInetAddresses.asScala.toSeq.filterNot { addr =>
+            addr.isLinkLocalAddress || addr.isLoopbackAddress
+          }
+          addr.nonEmpty
+        } match {
+          case Some(ni) =>
+            val addr = ni.getInetAddresses.asScala.toSeq.filterNot { inet =>
+              inet.isLinkLocalAddress || inet.isLoopbackAddress
+            }
+            val address = addr.find(_.isInstanceOf[Inet4Address]).getOrElse(addr.head).getAddress
+            // because of Inet6Address.toHostName may add interface at the end if it knows about it
+            InetAddress.getByAddress(address)
+          case None => address
+        }
+      case false => address
+    }
+  }
+
+  /** start service, if port is collision, retry 128 times */
+  def startServiceOnPort[T](
+    startPort: Int,
+    startService: Int => T,
+    maxRetries: Int = 128,
+    serviceName: String = ""): T = {
+
+    if (startPort != 0 && (startPort < 1024 || startPort > 65536)) {
+      throw new Exception("startPort should be between 1024 and 65535 (inclusive), " +
+        "or 0 for a random free port.")
+    }
+
+    val serviceString = if (serviceName.isEmpty) "" else s" '$serviceName'"
+    for (offset <- 0 to maxRetries) {
+      // Do not increment port if startPort is 0, which is treated as a special port
+      val tryPort = if (startPort == 0) {
+        startPort
+      } else {
+        // If the new port wraps around, do not try a privilege port
+        ((startPort + offset - 1024) % (65536 - 1024)) + 1024
+      }
+
+      try {
+        val result = startService(tryPort)
+        logger.info(s"Successfully started service$serviceString, result:$result.")
+        return result
+      } catch {
+        case e: Exception if isBindCollision(e) =>
+          if (offset >= maxRetries) {
+            val exceptionMessage = s"${e.getMessage}: Service$serviceString failed after " +
+              s"$maxRetries retries! Consider explicitly setting the appropriate port for the " +
+              s"service$serviceString (for example spark.ui.port for SparkUI) to an available " +
+              "port or increasing spark.port.maxRetries."
```

Comment:
@rmetzger I've added example to TcpReceiverSource and HttpReceiverSource. Also there are two full example in test package `TcpSourceExample` and `StreamSqlExample`